File: appinventor/components/src/com/google/appinventor/components/common/LOBFValues.java
// -*- mode: java; c-basic-offset: 2; -*-
// Copyright 2023-2024 MIT, All rights reserved
// Released under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

package com.google.appinventor.components.common;

import java.util.HashMap;
import java.util.Map;

/**
 * Defines a parameter of LOBF formula used by the regression components.
 */
@SuppressWarnings("checkstyle:AbbreviationAsWordInName")
public enum LOBFValues implements OptionList<String> {
  CorrCoef("correlation coefficient"),
  @Default
  Slope("slope"),
  Yintercept("Yintercept"),
  Predictions("predictions"),
  AllValues("all values"),
  QuadraticCoefficient("Quadratic Coefficient"),
  LinearCoefficient("slope"),
  ExponentialCoefficient("a"),
  ExponentialBase("b"),
  LogarithmCoefficient("b"),
  LogarithmConstant("a"),
  XIntercepts("Xintercepts"),
  RSquared("r^2");

  private final String lobfValues;

  LOBFValues(String lobfV) {
    this.lobfValues = lobfV;
  }

  public String toUnderlyingValue() {
    return lobfValues;
  }

  private static final Map<String, LOBFValues> lookup = new HashMap<>();

  static {
    for (LOBFValues value : LOBFValues.values()) {
      lookup.put(value.toUnderlyingValue(), value);
    }
  }

  public static LOBFValues fromUnderlyingValue(String value) {
    return lookup.get(value);
  }
}

File: appinventor/components-ios/src/LOBFValues.swift
// -*- mode: swift; swift-mode:basic-offset: 2; -*-
// Copyright 2025 MIT, All rights reserved
// Released under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

import Foundation

@objc public class LOBFValues: NSObject, OptionList {
  @objc public static let CorrCoef = LOBFValues("correlation coefficient")
  @objc public static let Slope = LOBFValues("slope")
  @objc public static let Yintercept = LOBFValues("Yintercept")
  @objc public static let Predictions = LOBFValues("predictions")
  @objc public static let AllValues = LOBFValues("all values")
  @objc public static let QuadraticCoefficient = LOBFValues("Quadratic Coefficient")
  @objc public static let LinearCoefficient = LOBFValues("slope")
  @objc public static let ExponentialCoefficient = LOBFValues("a")
  @objc public static let ExponentialBase = LOBFValues("b")
  @objc public static let LogarithmCoefficient = LOBFValues("b")
  @objc public static let LogarithmConstant = LOBFValues("a")
  @objc public static let XIntercepts = LOBFValues("XIntercepts")
  @objc public static let RSquared = LOBFValues("r^2")

  private static let LOOKUP: [String:LOBFValues] = generateOptionsLookup(
    CorrCoef,
    Slope,
    Yintercept,
    Predictions,
    AllValues,
    QuadraticCoefficient,
    LinearCoefficient,
    ExponentialCoefficient,
    ExponentialBase,
    LogarithmCoefficient,
    LogarithmConstant,
    XIntercepts,
    RSquared
  )

  let label: NSString

  @objc private init(_ label: NSString) {
    self.label = label
  }

  @objc class func fromUnderlyingValue(_ name: String) -> LOBFValues? {
    return LOOKUP[name]
  }

  @objc public func toUnderlyingValue() -> AnyObject {
    return label
  }
}
